atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -80,117 +66,143 @@ private CassandraTimestampBoundStore(CassandraClientPool clientPool) {
                     ConsistencyLevel.EACH_QUORUM);
+            return result;
         } catch (Exception e) {
-            log.error("[CAS] Error trying to set from {} to {}", oldVal, newVal, e);
-            DebugLogger.logger.error("[CAS] Error trying to set from {} to {}", oldVal, newVal, e);
-            throw Throwables.throwUncheckedException(e);
+            CassandraTimestampUtils.throwUpdateUncheckedException(entryInDb, newEntry, e);
+            return null;
         }
-        if (!result.isSuccess()) {
-            String msg = "Unable to CAS from {} to {}."
-                    + " Timestamp limit changed underneath us (limit in memory: {}, stored in DB: {})."
-                    + " This may indicate that another timestamp service is running against this cassandra keyspace."
-                    + " This is likely caused by multiple copies of a service running without a configured set of"
-                    + " leaders or a CLI being run with an embedded timestamp service against an already running"
-                    + " service.";
-            MultipleRunningTimestampServiceError err = new MultipleRunningTimestampServiceError(
-                    String.format(replaceBracesWithStringFormatSpecifier(msg),
-                            oldVal,
-                            newVal,
-                            currentLimit,
-                            getCurrentTimestampValues(result)));
-            log.error(msg, oldVal, newVal, currentLimit, getCurrentTimestampValues(result), err);
-            DebugLogger.logger.error(msg, oldVal, newVal, currentLimit, getCurrentTimestampValues(result), err);
-            DebugLogger.logger.error("Thread dump: {}", ThreadDumps.programmaticThreadDump());
-            throw err;
+    }
+
+    private void setCurrentLimit(String type, TimestampBoundStoreEntry newEntry) {
+        checkLimitNotDecreasing(newEntry);
+        currentLimit = newEntry.getTimestamp();
+        DebugLogger.logger.info("{} Setting cached timestamp limit to {}.", type, currentLimit);
+    }
+
+    private void retryCasIfMatchingId(Client client, TimestampBoundStoreEntry newEntry, CASResult result) {
+        TimestampBoundStoreEntry entryInDb = TimestampBoundStoreEntry.createFromCasResult(result);
+        if (entryInDb.idMatches(id)) {
+            setCurrentLimit("[CAS]", entryInDb);
+            entryInDb = TimestampBoundStoreEntry.create(currentLimit, id);
+            cas(client, entryInDb, newEntry);
         } else {
-            DebugLogger.logger.info("[CAS] Setting cached limit to {}.", newVal);
-            currentLimit = newVal;
+            CassandraTimestampUtils.throwStoringMultipleRunningTimestampServiceError(currentLimit, id,
+                    entryInDb, newEntry);
         }
     }
 
-    private String replaceBracesWithStringFormatSpecifier(String msg) {
-        return msg.replaceAll("\\{\\}", "%s");
+    private void checkValidId(TimestampBoundStoreEntry entryInDb) {
```

Comment:
nit: I'd prefer something like `matching` or `correct`; it would seem a different UUID should be "valid" in the validation sense, just wrong.